design/resource-management.md

```diff
@@ -0,0 +1,329 @@
+                       ├─{qemu-system-x86}(18061)
+                       ├─{qemu-system-x86}(18280)
+                       ├─{qemu-system-x86}(18281)
+                       ├─{qemu-system-x86}(18368)
+                       ├─{qemu-system-x86}(18369)
+                       └─{qemu-system-x86}(18370)
+{qemu-system-x86}(18059)
+vhost-18058(18060)
+```
+
+Two things should happen here:
+ * work with CRI-O to determine a more appropriate CPU shares setting for conmon, to avoid impacting the container
+ cgroups (in case of runc) or the hypervisor's cgroup (in case of Kata).  See <ADD ISSUE LINK HERE>
+ * do not place our IO threads, shim, proxy and QEMU process in conmon
+
+### Issues with current implementation
+
+There are a few major issues that result from the current implementation, and are motivation for design
+changes.
+
+#### Accurate usage accounting
+The IO pocessing should be charged to the pod performing the IO, not against the system. Without
+utilization of a same hierarchical cgroup, this will not be feasible.
+
+#### Node stability
+
+QEMU and its IO threads consume a non-negligible amount of resources. If the memory and CPU utilized is not
+constrained, measured and not accounted for, the node will run into CPU and memory pressure unexpectedly.
+
+#### Consistent guaranteed pod behavior
+
+Predictable performance is important for end users. By pushing IO threads into a shared pool, the
+achievable performance will be inconsistent.  Even if a user utilizes a `guaranteed` QoS pod, the
+performance profile will differ depending on the amount of contention on the system.  Raw unconstrained
+performance is important for Kata, but not as important as consistent and predictable behavior.
+
+#### OOM, unbound CPU utilization
+
+Memory limits are enforced, not requests. Until [Pod Overhead KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/20190226-pod-overhead.md)
+is added, users or admission controllers need to provide higher limits.
+
+In the case of CRI-O, the memory is being charged to conmon which is bounded by pod limits. As a result,
+the workload can still be OOMed. This is okay from a node stability point of view, but not from a pod stability
+point of view. This behavior assumes we are called by conmon and that they are constrained appropriately. Luckily,
+this is reasonably correct from a memory point of view. I/O bound workloads will exhibit sub-optimal performance
+due to the CPU constraints applied to conmon (where the io threads run).
+
+For containerd, the memory is being charged to containerd, which is basically unbounded.  This is bad for
+node stability, as the `pod` is essentially unbounded.
+
+## Proposed Changes
+
+### Summary
+ * Pause cgroup cpu shaes should be setup correctly.
+ * Do not create container cgroups on the host. Instead, create a pod sandbox group that is entirely managed by Kata
+ * Move the QEMU threads into the sandbox cgroup
```

Comment:
Plus shimv2, kata-shim/proxy, and vhost threads? I think they all belong to the sandbox cgroup.